`dqrng` 0.4.0 fails to build: `error: no match for 'operator/'`

```
--->  Building R-dqrng
xinstall: mkdir /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/dqrng" && /opt/local/bin/R CMD INSTALL . --library=/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build --install-tests 
* installing *source* package ‘dqrng’ ...
** package ‘dqrng’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
using C++ compiler: ‘g++-mp-13 (MacPorts gcc13 13.2.0_4+stdlib_flag) 13.2.0’
Warning in system2("xcrun", "--show-sdk-path", TRUE, TRUE) :
  running command ''xcrun' --show-sdk-path 2>&1' had status 64
using SDK: ‘NA’‘NA’‘NA’‘NA’‘NA’‘NA’
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c RcppExports.cpp -o RcppExports.o
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c dqrng.cpp -o dqrng.o
In file included from dqrng.cpp:21:
../inst/include/dqrng_generator.h: In member function 'void dqrng::random_64bit_wrapper<RNG>::set_stream(dqrng::random_64bit_generator::result_type) [with RNG = pcg_detail::engine<long long unsigned int, pcg_extras::uint_x4<unsigned int, long long unsigned int>, pcg_detail::xsl_rr_mixin<long long unsigned int, pcg_extras::uint_x4<unsigned int, long long unsigned int> >, false, pcg_detail::specific_stream<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_detail::default_multiplier<pcg_extras::uint_x4<unsigned int, long long unsigned int> > >; dqrng::random_64bit_generator::result_type = long long unsigned int]':
../inst/include/dqrng_generator.h:115:26: error: no match for 'operator/' (operand types are '__gnu_cxx::__alloc_traits<std::allocator<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_extras::uint_x4<unsigned int, long long unsigned int> >::value_type' {aka 'pcg_extras::uint_x4<unsigned int, long long unsigned int>'} and 'int')
  115 |   gen.set_stream(state[1]/2 + stream);
../inst/include/dqrng_generator.h:115:26: note: candidate: 'operator/(int, int)' (built-in)
../inst/include/dqrng_generator.h:115:26: note:   no known conversion for argument 1 from '__gnu_cxx::__alloc_traits<std::allocator<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_extras::uint_x4<unsigned int, long long unsigned int> >::value_type' {aka 'pcg_extras::uint_x4<unsigned int, long long unsigned int>'} to 'int'
../inst/include/dqrng_generator.h:115:26: note: candidate: 'operator/(long unsigned int, int)' (built-in)
../inst/include/dqrng_generator.h:115:26: note:   no known conversion for argument 1 from '__gnu_cxx::__alloc_traits<std::allocator<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_extras::uint_x4<unsigned int, long long unsigned int> >::value_type' {aka 'pcg_extras::uint_x4<unsigned int, long long unsigned int>'} to 'long unsigned int'
../inst/include/dqrng_generator.h:115:26: note: candidate: 'operator/(unsigned int, int)' (built-in)
../inst/include/dqrng_generator.h:115:26: note:   no known conversion for argument 1 from '__gnu_cxx::__alloc_traits<std::allocator<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_extras::uint_x4<unsigned int, long long unsigned int> >::value_type' {aka 'pcg_extras::uint_x4<unsigned int, long long unsigned int>'} to 'unsigned int'
../inst/include/dqrng_generator.h:115:26: note: candidate: 'operator/(long long unsigned int, int)' (built-in)
../inst/include/dqrng_generator.h:115:26: note:   no known conversion for argument 1 from '__gnu_cxx::__alloc_traits<std::allocator<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_extras::uint_x4<unsigned int, long long unsigned int> >::value_type' {aka 'pcg_extras::uint_x4<unsigned int, long long unsigned int>'} to 'long long unsigned int'
In file included from ../inst/include/pcg_extras.hpp:84,
                 from ../inst/include/dqrng_types.h:27,
                 from ../inst/include/dqrng_generator.h:24:
../inst/include/pcg_uint128.hpp:559:22: note: candidate: 'template<class UInt, class UIntX2> pcg_extras::uint_x4<U, V> pcg_extras::operator/(const uint_x4<U, V>&, const uint_x4<U, V>&)'
  559 | uint_x4<UInt,UIntX2> operator/(const uint_x4<UInt,UIntX2>& dividend,
      |                      ^~~~~~~~
../inst/include/pcg_uint128.hpp:559:22: note:   template argument deduction/substitution failed:
../inst/include/dqrng_generator.h:115:27: note:   mismatched types 'const pcg_extras::uint_x4<U, V>' and 'int'
  115 |   gen.set_stream(state[1]/2 + stream);
      |                           ^
In file included from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/RcppCommon.h:137,
                 from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp.h:27,
                 from dqrng.cpp:20:
/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/complex.h:46:17: note: candidate: 'Rcomplex operator/(const Rcomplex&, const Rcomplex&)'
   46 | inline Rcomplex operator/( const Rcomplex& a, const Rcomplex& b) {
      |                 ^~~~~~~~
/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/complex.h:46:44: note:   no known conversion for argument 1 from '__gnu_cxx::__alloc_traits<std::allocator<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_extras::uint_x4<unsigned int, long long unsigned int> >::value_type' {aka 'pcg_extras::uint_x4<unsigned int, long long unsigned int>'} to 'const Rcomplex&'
   46 | inline Rcomplex operator/( const Rcomplex& a, const Rcomplex& b) {
      |                            ~~~~~~~~~~~~~~~~^
In file included from ../inst/include/dqrng_generator.h:26:
../inst/include/pcg_random.hpp: In instantiation of 'void pcg_detail::specific_stream<itype>::set_stream(itype) [with itype = pcg_extras::uint_x4<unsigned int, long long unsigned int>]':
../inst/include/pcg_random.hpp:610:26:   required from 'std::basic_istream<_CharT, _Traits>& pcg_detail::operator>>(std::basic_istream<_CharT, _Traits>&, engine<xtype, itype, output_mixin, output_previous, stream_mixin, multiplier_mixin>&) [with CharT = char; Traits = std::char_traits<char>; xtype = long long unsigned int; itype = pcg_extras::uint_x4<unsigned int, long long unsigned int>; output_mixin = xsl_rr_mixin<long long unsigned int, pcg_extras::uint_x4<unsigned int, long long unsigned int> >; bool output_previous = false; stream_mixin = specific_stream<pcg_extras::uint_x4<unsigned int, long long unsigned int> >; multiplier_mixin = default_multiplier<pcg_extras::uint_x4<unsigned int, long long unsigned int> >]'
../inst/include/dqrng_generator.h:55:55:   required from 'void dqrng::random_64bit_wrapper<RNG>::input(std::istream&) [with RNG = pcg_detail::engine<long long unsigned int, pcg_extras::uint_x4<unsigned int, long long unsigned int>, pcg_detail::xsl_rr_mixin<long long unsigned int, pcg_extras::uint_x4<unsigned int, long long unsigned int> >, false, pcg_detail::specific_stream<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_detail::default_multiplier<pcg_extras::uint_x4<unsigned int, long long unsigned int> > >; std::istream = std::basic_istream<char>]'
../inst/include/dqrng_generator.h:55:16:   required from here
../inst/include/pcg_random.hpp:322:37: error: no match for 'operator|' (operand types are 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' and 'int')
  322 |          inc_ = (specific_seq << 1) | 1;
      |                 ~~~~~~~~~~~~~~~~~~~~^~~
../inst/include/pcg_random.hpp:322:37: note: candidate: 'operator|(int, int)' (built-in)
../inst/include/pcg_random.hpp:322:37: note:   no known conversion for argument 1 from 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' to 'int'
../inst/include/pcg_random.hpp:322:37: note: candidate: 'operator|(long unsigned int, int)' (built-in)
../inst/include/pcg_random.hpp:322:37: note:   no known conversion for argument 1 from 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' to 'long unsigned int'
../inst/include/pcg_random.hpp:322:37: note: candidate: 'operator|(unsigned int, int)' (built-in)
../inst/include/pcg_random.hpp:322:37: note:   no known conversion for argument 1 from 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' to 'unsigned int'
../inst/include/pcg_random.hpp:322:37: note: candidate: 'operator|(long long unsigned int, int)' (built-in)
../inst/include/pcg_random.hpp:322:37: note:   no known conversion for argument 1 from 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' to 'long long unsigned int'
../inst/include/pcg_uint128.hpp:664:22: note: candidate: 'template<class UInt, class UIntX2> pcg_extras::uint_x4<U, V> pcg_extras::operator|(const uint_x4<U, V>&, const uint_x4<U, V>&)'
  664 | uint_x4<UInt,UIntX2> operator|(const uint_x4<UInt,UIntX2>& a,
      |                      ^~~~~~~~
../inst/include/pcg_uint128.hpp:664:22: note:   template argument deduction/substitution failed:
../inst/include/pcg_random.hpp:322:37: note:   mismatched types 'const pcg_extras::uint_x4<U, V>' and 'int'
  322 |          inc_ = (specific_seq << 1) | 1;
      |                 ~~~~~~~~~~~~~~~~~~~~^~~
In file included from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/sugar/logical/logical.h:30,
                 from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/sugar/sugar_forward.h:29,
                 from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/RcppCommon.h:184:
/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/sugar/logical/or.h:315:1: note: candidate: 'template<bool LHS_NA, class LHS_T, bool RHS_NA, class RHS_T> Rcpp::sugar::Or_LogicalExpression_LogicalExpression<LHS_NA, LHS_T, RHS_NA, RHS_T> operator|(const Rcpp::VectorBase<10, LHS_NA, LHS_T>&, const Rcpp::VectorBase<10, RHS_NA, RHS_T>&)'
  315 | operator|(
      | ^~~~~~~~
/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/sugar/logical/or.h:315:1: note:   template argument deduction/substitution failed:
../inst/include/pcg_random.hpp:322:37: note:   'pcg_extras::uint_x4<unsigned int, long long unsigned int>' is not derived from 'const Rcpp::VectorBase<10, LHS_NA, LHS_T>'
  322 |          inc_ = (specific_seq << 1) | 1;
      |                 ~~~~~~~~~~~~~~~~~~~~^~~
make: *** [dqrng.o] Error 1
ERROR: compilation failed for package ‘dqrng’
```

Looks like the last non-CRAN version to build is 0.3.1.2, i.e. ad154c2dd1b724a7306b3e2d601040afefd6836a commit:
```
--->  Building R-dqrng
xinstall: mkdir /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/dqrng-ad154c2dd1b724a7306b3e2d601040afefd6836a" && /opt/local/bin/R CMD INSTALL . --library=/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build --install-tests 
* installing *source* package ‘dqrng’ ...
** using staged installation
** libs
using C++ compiler: ‘g++-mp-13 (MacPorts gcc13 13.2.0_4+stdlib_flag) 13.2.0’
Warning in system2("xcrun", "--show-sdk-path", TRUE, TRUE) :
  running command ''xcrun' --show-sdk-path 2>&1' had status 64
using SDK: ‘NA’‘NA’‘NA’‘NA’‘NA’‘NA’
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c RcppExports.cpp -o RcppExports.o
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c dqrng.cpp -o dqrng.o
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c generateSeedVectors.cpp -o generateSeedVectors.o
/opt/local/bin/g++-mp-13 -std=gnu++17 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/opt/local/Library/Frameworks/R.framework/Resources/lib -Wl,-headerpad_max_install_names -L/opt/local/lib -lMacportsLegacySupport -Wl,-rpath,/opt/local/lib/libgcc -arch ppc -o dqrng.so RcppExports.o dqrng.o generateSeedVectors.o -F/opt/local/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
installing to /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build/00LOCK-dqrng-ad154c2dd1b724a7306b3e2d601040afefd6836a/00new/dqrng/libs
** R
** inst
** tests
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (dqrng)
```

0.3.1.3, i.e. commit abdb3ebba26dc78f0f0502c6c10ab894ae46a30e – is already broken:
```
--->  Building R-dqrng
xinstall: mkdir /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/dqrng-abdb3ebba26dc78f0f0502c6c10ab894ae46a30e" && /opt/local/bin/R CMD INSTALL . --library=/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build --install-tests 
* installing *source* package ‘dqrng’ ...
** using staged installation
** libs
using C++ compiler: ‘g++-mp-13 (MacPorts gcc13 13.2.0_4+stdlib_flag) 13.2.0’
Warning in system2("xcrun", "--show-sdk-path", TRUE, TRUE) :
  running command ''xcrun' --show-sdk-path 2>&1' had status 64
using SDK: ‘NA’‘NA’‘NA’‘NA’‘NA’‘NA’
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c RcppExports.cpp -o RcppExports.o
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c dqrng.cpp -o dqrng.o
In file included from ../inst/include/dqrng_generator.h:25,
                 from dqrng.cpp:21:
../inst/include/pcg_random.hpp: In instantiation of 'void pcg_detail::specific_stream<itype>::set_stream(itype) [with itype = pcg_extras::uint_x4<unsigned int, long long unsigned int>]':
../inst/include/pcg_random.hpp:610:26:   required from 'std::basic_istream<_CharT, _Traits>& pcg_detail::operator>>(std::basic_istream<_CharT, _Traits>&, engine<xtype, itype, output_mixin, output_previous, stream_mixin, multiplier_mixin>&) [with CharT = char; Traits = std::char_traits<char>; xtype = long long unsigned int; itype = pcg_extras::uint_x4<unsigned int, long long unsigned int>; output_mixin = xsl_rr_mixin<long long unsigned int, pcg_extras::uint_x4<unsigned int, long long unsigned int> >; bool output_previous = false; stream_mixin = specific_stream<pcg_extras::uint_x4<unsigned int, long long unsigned int> >; multiplier_mixin = default_multiplier<pcg_extras::uint_x4<unsigned int, long long unsigned int> >]'
../inst/include/dqrng_generator.h:56:55:   required from 'void dqrng::random_64bit_wrapper<RNG>::input(std::istream&) [with RNG = pcg_detail::engine<long long unsigned int, pcg_extras::uint_x4<unsigned int, long long unsigned int>, pcg_detail::xsl_rr_mixin<long long unsigned int, pcg_extras::uint_x4<unsigned int, long long unsigned int> >, false, pcg_detail::specific_stream<pcg_extras::uint_x4<unsigned int, long long unsigned int> >, pcg_detail::default_multiplier<pcg_extras::uint_x4<unsigned int, long long unsigned int> > >; std::istream = std::basic_istream<char>]'
../inst/include/dqrng_generator.h:56:16:   required from here
../inst/include/pcg_random.hpp:322:37: error: no match for 'operator|' (operand types are 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' and 'int')
  322 |          inc_ = (specific_seq << 1) | 1;
      |                 ~~~~~~~~~~~~~~~~~~~~^~~
../inst/include/pcg_random.hpp:322:37: note: candidate: 'operator|(int, int)' (built-in)
../inst/include/pcg_random.hpp:322:37: note:   no known conversion for argument 1 from 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' to 'int'
../inst/include/pcg_random.hpp:322:37: note: candidate: 'operator|(long unsigned int, int)' (built-in)
../inst/include/pcg_random.hpp:322:37: note:   no known conversion for argument 1 from 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' to 'long unsigned int'
../inst/include/pcg_random.hpp:322:37: note: candidate: 'operator|(unsigned int, int)' (built-in)
../inst/include/pcg_random.hpp:322:37: note:   no known conversion for argument 1 from 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' to 'unsigned int'
../inst/include/pcg_random.hpp:322:37: note: candidate: 'operator|(long long unsigned int, int)' (built-in)
../inst/include/pcg_random.hpp:322:37: note:   no known conversion for argument 1 from 'pcg_extras::uint_x4<unsigned int, long long unsigned int>' to 'long long unsigned int'
In file included from ../inst/include/pcg_extras.hpp:84,
                 from ../inst/include/pcg_random.hpp:114:
../inst/include/pcg_uint128.hpp:664:22: note: candidate: 'template<class UInt, class UIntX2> pcg_extras::uint_x4<U, V> pcg_extras::operator|(const uint_x4<U, V>&, const uint_x4<U, V>&)'
  664 | uint_x4<UInt,UIntX2> operator|(const uint_x4<UInt,UIntX2>& a,
      |                      ^~~~~~~~
../inst/include/pcg_uint128.hpp:664:22: note:   template argument deduction/substitution failed:
../inst/include/pcg_random.hpp:322:37: note:   mismatched types 'const pcg_extras::uint_x4<U, V>' and 'int'
  322 |          inc_ = (specific_seq << 1) | 1;
      |                 ~~~~~~~~~~~~~~~~~~~~^~~
In file included from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/sugar/logical/logical.h:30,
                 from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/sugar/sugar_forward.h:29,
                 from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/RcppCommon.h:184,
                 from /opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp.h:27,
                 from dqrng.cpp:20:
/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/sugar/logical/or.h:315:1: note: candidate: 'template<bool LHS_NA, class LHS_T, bool RHS_NA, class RHS_T> Rcpp::sugar::Or_LogicalExpression_LogicalExpression<LHS_NA, LHS_T, RHS_NA, RHS_T> operator|(const Rcpp::VectorBase<10, LHS_NA, LHS_T>&, const Rcpp::VectorBase<10, RHS_NA, RHS_T>&)'
  315 | operator|(
      | ^~~~~~~~
/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include/Rcpp/sugar/logical/or.h:315:1: note:   template argument deduction/substitution failed:
../inst/include/pcg_random.hpp:322:37: note:   'pcg_extras::uint_x4<unsigned int, long long unsigned int>' is not derived from 'const Rcpp::VectorBase<10, LHS_NA, LHS_T>'
  322 |          inc_ = (specific_seq << 1) | 1;
      |                 ~~~~~~~~~~~~~~~~~~~~^~~
make: *** [dqrng.o] Error 1
ERROR: compilation failed for package ‘dqrng’
* removing ‘/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build/dqrng’
Command failed:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/dqrng-abdb3ebba26dc78f0f0502c6c10ab894ae46a30e" && /opt/local/bin/R CMD INSTALL . --library=/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build --install-tests 
Exit code: 1
Error: Failed to build R-dqrng: command execution failed
```

However whatever CRAN has as 0.3.2 builds with no issues.

@rstub @daqana Could someone please help me here?

@rstub The build from fd688fb4c28ac8d36b9b7cf5ce689648eb5f66d7 works, though with some warnings:
```
--->  Building R-dqrng
xinstall: mkdir /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/dqrng-fd688fb4c28ac8d36b9b7cf5ce689648eb5f66d7" && /opt/local/bin/R CMD INSTALL . --library=/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build --install-tests 
* installing *source* package ‘dqrng’ ...
** using staged installation
** libs
using C++ compiler: ‘g++-mp-13 (MacPorts gcc13 13.2.0_4+stdlib_flag) 13.2.0’
Warning in system2("xcrun", "--show-sdk-path", TRUE, TRUE) :
  running command ''xcrun' --show-sdk-path 2>&1' had status 64
using SDK: ‘NA’‘NA’‘NA’‘NA’‘NA’‘NA’
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c RcppExports.cpp -o RcppExports.o
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c dqrng.cpp -o dqrng.o
In file included from ../inst/include/dqrng_generator.h:25,
                 from dqrng.cpp:21:
In member function 'void dqrng::xoshiro<N>::do_jump(std::array<long long unsigned int, N>) [with long unsigned int N = 2]',
    inlined from 'void dqrng::xoshiro<N>::jump() [with long unsigned int N = 2]' at ../inst/include/xoshiro.h:75:12:
../inst/include/xoshiro.h:112:21: warning: iteration 2 invokes undefined behavior [-Waggressive-loop-optimizations]
  112 |         if (JUMP[i] & UINT64_C(1) << b) {
../inst/include/xoshiro.h:110:31: note: within this loop
  110 |     for(unsigned int i = 0; i < sizeof JUMP / sizeof JUMP.begin(); i++)
      |                             ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In member function 'void dqrng::xoshiro<N>::do_jump(std::array<long long unsigned int, N>) [with long unsigned int N = 4]',
    inlined from 'void dqrng::xoshiro<N>::long_jump() [with long unsigned int N = 4]' at ../inst/include/xoshiro.h:81:12:
../inst/include/xoshiro.h:131:21: warning: iteration 4 invokes undefined behavior [-Waggressive-loop-optimizations]
  131 |         if (JUMP[i] & UINT64_C(1) << b) {
../inst/include/xoshiro.h:129:31: note: within this loop
  129 |     for(unsigned int i = 0; i < sizeof JUMP / sizeof JUMP.begin(); i++)
      |                             ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I../inst/include -DSTRICT_R_HEADERS -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/BH/include' -I'/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/sitmo/include' -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c generateSeedVectors.cpp -o generateSeedVectors.o
/opt/local/bin/g++-mp-13 -std=gnu++17 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/opt/local/Library/Frameworks/R.framework/Resources/lib -Wl,-headerpad_max_install_names -L/opt/local/lib -lMacportsLegacySupport -Wl,-rpath,/opt/local/lib/libgcc -arch ppc -o dqrng.so RcppExports.o dqrng.o generateSeedVectors.o -F/opt/local/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
installing to /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_R_R-dqrng/R-dqrng/work/build/00LOCK-dqrng-fd688fb4c28ac8d36b9b7cf5ce689648eb5f66d7/00new/dqrng/libs
** R
```
And this:
```
* checking compiled code ... WARNING
File ‘dqrng/libs/dqrng.so’:
  Found ‘_abort’, possibly from ‘abort’ (C)
    Object: ‘dqrng.o’

Compiled code should not call entry points which might terminate R nor
write to stdout/stderr instead of to the console, nor use Fortran I/O
nor system RNGs nor [v]sprintf.
```
There is a failure in tests now, though it may not be in fact related to `dqrng` at all:
```
* checking tests ...
  Running ‘testthat.R’/opt/local/Library/Frameworks/R.framework/Resources/bin/BATCH: line 60: 15027 Bus error               ${R_HOME}/bin/R -f ${in} ${opts} ${R_BATCH_OPTIONS} > ${out} 2>&1

 ERROR
Running the tests in ‘tests/testthat.R’ failed.
Last 13 lines of output:
  26: tryCatch(withCallingHandlers({    eval(code, test_env)    if (!handled && !is.null(test)) {        skip_empty()    }}, expectation = handle_expectation, skip = handle_skip, warning = handle_warning,     message = handle_message, error = handle_error), error = handle_fatal,     skip = function(e) {    })
  27: test_code(test = NULL, code = exprs, env = env, default_reporter = StopReporter$new())
  28: source_file(path, env = env(env), desc = desc, error_call = error_call)
  29: FUN(X[[i]], ...)
  30: lapply(test_paths, test_one_file, env = env, desc = desc, error_call = error_call)
  31: doTryCatch(return(expr), name, parentenv, handler)
  32: tryCatchOne(expr, names, parentenv, handlers[[1L]])
  33: tryCatchList(expr, classes, parentenv, handlers)
  34: tryCatch(code, testthat_abort_reporter = function(cnd) {    cat(conditionMessage(cnd), "\n")    NULL})
  35: with_reporter(reporters$multi, lapply(test_paths, test_one_file,     env = env, desc = desc, error_call = error_call))
  36: test_files_serial(test_dir = test_dir, test_package = test_package,     test_paths = test_paths, load_helpers = load_helpers, reporter = reporter,     env = env, stop_on_failure = stop_on_failure, stop_on_warning = stop_on_warning,     desc = desc, load_package = load_package, error_call = error_call)
  37: test_files(test_dir = path, test_paths = test_paths, test_package = package,     reporter = reporter, load_helpers = load_helpers, env = env,     stop_on_failure = stop_on_failure, stop_on_warning = stop_on_warning,     load_package = load_package, parallel = parallel)
  38: test_dir("testthat", package = package, reporter = reporter,     ..., load_package = "installed")
  39: test_check("dqrng")
  An irrecoverable exception occurred. R is aborting now ...
* checking for unstated dependencies in vignettes ... OK
* checking package vignettes ... WARNING
Directory 'inst/doc' does not exist.
Package vignettes without corresponding single PDF/HTML:
  ‘cpp-api.Rmd’
  ‘dqrng.Rmd’
  ‘parallel.Rmd’
  ‘sample.Rmd’
* checking running R code from vignettes ...
  ‘cpp-api.Rmd’ using ‘UTF-8’... OK
  ‘dqrng.Rmd’ using ‘UTF-8’... OK
  ‘parallel.Rmd’ using ‘UTF-8’... OK
  ‘sample.Rmd’ using ‘UTF-8’... OK
 OK
* checking re-building of vignette outputs ... SKIPPED
* DONE

Status: 1 ERROR, 3 WARNINGs
```
It looks very similar to this bug in `R`: https://trac.macports.org/ticket/67059
I have no idea what is that and why it causes Bus error, but it seems it is in `R` itself.

Do I need some specific version?
```
36-244% Rscript -e "Rcpp::sourceCpp('test_dqrng2.cpp')"
using C++ compiler: ‘g++-mp-13 (MacPorts gcc13 13.2.0_4+stdlib_flag) 13.2.0’
using SDK: ‘NA’‘NA’‘NA’‘NA’‘NA’‘NA’
/opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I"/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include" -I"/Users/svacchanda/Desktop/_opt_PPCSnowLeopardPorts_lang_guile2/test" -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c test_dqrng2.cpp -o test_dqrng2.o
test_dqrng2.cpp:4:10: fatal error: pcg_random.hpp: No such file or directory
    4 | #include <pcg_random.hpp>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
make: *** [test_dqrng2.o] Error 1
Warning message:
In system2("xcrun", "--show-sdk-path", TRUE, TRUE) :
  running command ''xcrun' --show-sdk-path 2>&1' had status 64
Error in Rcpp::sourceCpp("test_dqrng2.cpp") : 
  Error 1 occurred building shared library.
Execution halted
```
I do not know where this header is supposed to come from. (But since I build this manually and not in MacPorts, environment may not know about other R packages, unless those are explicitly specified.)

Sorry for being unclear. This should be compiled w/o using R at all, i.e.
`g++ -I<path/to/include> -o dqrng_test2 dqrng_trst2.cpp`.


Sergey Fedorov ***@***.***> schrieb am Mo. 20. Mai 2024 um
12:06:

> Do I need some specific version?
>
> 36-244% Rscript -e "Rcpp::sourceCpp('test_dqrng2.cpp')"
> using C++ compiler: ‘g++-mp-13 (MacPorts gcc13 13.2.0_4+stdlib_flag) 13.2.0’
> using SDK: ‘NA’‘NA’‘NA’‘NA’‘NA’‘NA’
> /opt/local/bin/g++-mp-13 -std=gnu++17 -I"/opt/local/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I"/opt/local/Library/Frameworks/R.framework/Versions/4.4/Resources/library/Rcpp/include" -I"/Users/svacchanda/Desktop/_opt_PPCSnowLeopardPorts_lang_guile2/test" -isystem/opt/local/include/LegacySupport -I/opt/local/include    -fPIC  -pipe -Os -arch ppc   -c test_dqrng2.cpp -o test_dqrng2.o
> test_dqrng2.cpp:4:10: fatal error: pcg_random.hpp: No such file or directory
>     4 | #include <pcg_random.hpp>
>       |          ^~~~~~~~~~~~~~~~
> compilation terminated.
> make: *** [test_dqrng2.o] Error 1
> Warning message:
> In system2("xcrun", "--show-sdk-path", TRUE, TRUE) :
>   running command ''xcrun' --show-sdk-path 2>&1' had status 64
> Error in Rcpp::sourceCpp("test_dqrng2.cpp") :
>   Error 1 occurred building shared library.
> Execution halted
>
> I do not know where this header is supposed to come from. (But since I
> build this manually and not in MacPorts, environment may not know about
> other R packages, unless those are explicitly specified.)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/daqana/dqrng/issues/88#issuecomment-2120119464>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGUIIS2ZADMEE52GVRZ6XL3ZDHDJZAVCNFSM6AAAAABHZFYAZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRQGEYTSNBWGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

--- a/DESCRIPTION
+++ b/DESCRIPTION
@@ -1,7 +1,7 @@
 Package: dqrng
 Type: Package
 Title: Fast Pseudo Random Number Generators
-Version: 0.4.0.1
+Version: 0.4.0.2
 Authors@R: c(
     person("Ralf", "Stubner", email = "ralf.stubner@gmail.com", role = c("aut", "cre"), comment = c(ORCID = "0009-0009-1908-106X")),
     person("daqana GmbH", role = "cph"),

--- a/NEWS.md
+++ b/NEWS.md
@@ -2,6 +2,7 @@
 
 * Fix an UBSAN error found by CRAN ([#90](https://github.com/daqana/dqrng/pull/90) fixing [#89](https://github.com/daqana/dqrng/issues/89))
 * Fix a compilation error when the compiler does not provide a 128bit integer type ([#90](https://github.com/daqana/dqrng/pull/90) fixing [#88](https://github.com/daqana/dqrng/issues/88))
+* Disable PCG64 on MacOS PowerPC ([#91](https://github.com/daqana/dqrng/pull/91) fixing [#88](https://github.com/daqana/dqrng/issues/88))
 
 # dqrng 0.4.0
 

--- a/inst/include/dqrng_generator.h
+++ b/inst/include/dqrng_generator.h
@@ -100,6 +100,7 @@ inline void random_64bit_wrapper<::dqrng::xoshiro256starstar>::set_stream(result
   gen.long_jump(stream);
 }
 
+#if !(defined(__APPLE__) && defined(__POWERPC__))
 template<>
 inline void random_64bit_wrapper<pcg64>::set_stream(result_type stream) {
   // set the stream relative to the current stream, i.e. stream = 0 does not change the RNG
@@ -121,6 +122,7 @@ inline void random_64bit_wrapper<pcg64>::seed(result_type seed, result_type stre
   gen.seed(seed, stream);
   cache = false;
 }
+#endif
 
 inline uint64_t get_seed_from_r() {
   Rcpp::RNGScope rngScope;

--- a/inst/include/xoshiro.h
+++ b/inst/include/xoshiro.h
@@ -16,6 +16,7 @@
 #include <mystdint.h>
 #include <functional>
 #include <algorithm>
+#include <iostream>
 
 namespace dqrng {
 template<size_t N>

--- a/inst/include/xoshiro.h
+++ b/inst/include/xoshiro.h
@@ -108,7 +108,7 @@ template<>
 inline void xoshiro<2>::do_jump(std::array<result_type, 2> JUMP) {
     uint64_t s0 = 0;
     uint64_t s1 = 0;
-    for(unsigned int i = 0; i < sizeof JUMP / sizeof JUMP.begin(); i++)
+    for(int i = 0; i < 2; i++)
       for(int b = 0; b < 64; b++) {
         if (JUMP[i] & UINT64_C(1) << b) {
           s0 ^= s[0];
@@ -127,7 +127,7 @@ inline void xoshiro<4>::do_jump(std::array<result_type, 4> JUMP) {
     uint64_t s1 = 0;
     uint64_t s2 = 0;
     uint64_t s3 = 0;
-    for(unsigned int i = 0; i < sizeof JUMP / sizeof JUMP.begin(); i++)
+    for(int i = 0; i < 4; i++)
       for(int b = 0; b < 64; b++) {
         if (JUMP[i] & UINT64_C(1) << b) {
           s0 ^= s[0];

--- a/src/dqrng.cpp
+++ b/src/dqrng.cpp
@@ -66,8 +66,10 @@ void dqRNGkind(std::string kind, const std::string& normal_kind = "ignored") {
     rng =  dqrng::generator<dqrng::xoshiro256plus>(seed);
   } else if (kind == "xoshiro256++") {
     rng =  dqrng::generator<dqrng::xoshiro256plusplus>(seed);
+#if !(defined(__APPLE__) && defined(__POWERPC__))
   } else if (kind == "pcg64") {
     rng =  dqrng::generator<pcg64>(seed);
+#endif
   } else if (kind == "threefry") {
     rng =  dqrng::generator<sitmo::threefry_20_64>(seed);
   } else {

--- a/tests/testthat/helper.R
+++ b/tests/testthat/helper.R
@@ -1,0 +1,10 @@
+powerpc_apple <- grepl("powerpc-apple-", R.version$platform)
+
+safe_expect_error <- function(..., msg) {
+  os.type <- Sys.info()["sysname"]
+  if (os.type == "Darwin") {
+    expect_error(...)
+  } else {
+    expect_error(..., msg)
+  }
+}

--- a/tests/testthat/helper.R
+++ b/tests/testthat/helper.R
@@ -1,4 +1,4 @@
-powerpc_apple <- grepl("powerpc-apple-", R.version$platform)
+powerpc_apple <- grepl("powerpc-apple-darwin", R.version$platform)
 
 safe_expect_error <- function(..., msg) {
   os.type <- Sys.info()["sysname"]

--- a/tests/testthat/test-convert.R
+++ b/tests/testthat/test-convert.R
@@ -1,14 +1,5 @@
 context("conversion")
 
-safe_expect_error <- function(..., msg) {
-  os.type <- Sys.info()["sysname"]
-  if (os.type == "Darwin") {
-    expect_error(...)
-  } else {
-    expect_error(..., msg)
-  }
-}
-
 Rcpp::sourceCpp("cpp/convert.cpp") # Warnings about shifts can be ignored.
 
 test_that("conversion to 16-bit integers works correctly", {

--- a/tests/testthat/test-external-generator.R
+++ b/tests/testthat/test-external-generator.R
@@ -70,6 +70,7 @@ test_that("cloned external Xoshiro256++ gives different result only when a diffe
 })
 
 test_that("cloned external PCG64 gives different result", {
+  skip_if(powerpc_apple)
   dqrng::dqRNGkind("PCG64")
   dqset.seed(use_seed)
   expect_true(cloned_calls(stream = 0))

--- a/tests/testthat/test-generators.R
+++ b/tests/testthat/test-generators.R
@@ -40,6 +40,7 @@ test_that("Default generator: setting same seed but different stream produces di
 })
 
 test_that("PCG64: setting seed produces identical uniformly distributed numbers", {
+  skip_if(powerpc_apple)
   dqRNGkind("pcg64")
   dqset.seed(seed)
   u1 <- dqrunif(10)
@@ -49,6 +50,7 @@ test_that("PCG64: setting seed produces identical uniformly distributed numbers"
 })
 
 test_that("PCG64: setting seed and stream produces identical uniformly distributed numbers", {
+  skip_if(powerpc_apple)
   dqRNGkind("pcg64")
   dqset.seed(seed, 1)
   u1 <- dqrunif(10)
@@ -58,6 +60,7 @@ test_that("PCG64: setting seed and stream produces identical uniformly distribut
 })
 
 test_that("PCG64: saving state produces identical uniformly distributed numbers", {
+  skip_if(powerpc_apple)
   dqRNGkind("pcg64")
   dqset.seed(seed, 1)
   state <- dqrng_get_state()
@@ -68,6 +71,7 @@ test_that("PCG64: saving state produces identical uniformly distributed numbers"
 })
 
 test_that("PCG64: setting same seed but different stream produces different uniformly distributed numbers", {
+  skip_if(powerpc_apple)
   dqRNGkind("pcg64")
   dqset.seed(seed, 1)
   u1 <- dqrunif(10)

